### DISCUSSION TITLE
Practical guidance for installation of Readium on CentOS to serve a library of ePub documents

### DISCUSSION THREAD

**Discussion by Nainoia-Inc-Admin** [readium/architecture#156] [Q&A]
I maintain a library of ePub documents that are available for simple download and also for viewing online.  I provide these services from a CentOS web server and currently use Futurepress to allow online reading of the ePubs from my server.  Futurepress works well, but I was directed to Readium because it may be more well maintained with more features and a larger community of support.

Question, can someone direct me to documentation that explains which repositories I need and how to unpack and configure them onto my CentOS webserver so that I could possibly use Readium to allow online reading of my ePub library?

Thanks for any help,

Jeff

**Comment by danielweck** [comment]
Hello!
Well, there's the legacy Readium "1" ecosystem (in contrast to the so-codenamed Readium "2" project, which effectively consists in a wholesale refactoring based on an improved, more modern architecture).

Readium "1" provided a rendering engine, as well as an off-the-shelf online reader (client-side web browser app):
https://github.com/readium/readium-js-viewer

There are several Readium "2" cloud reader implementations, but the goal is to provide a consistent model (work in progress):
https://github.com/readium/web

So, at this point in time, there is no off-the-shelf solution under `/readium` ownership here in GitHub, but there are available spin-offs (I will let others chime in should they wish to provide guidance).

**Reply by Nainoia-Inc-Admin** [reply]
Okay trying to understand what you are saying.  So https://github.com/readium/readium-js-viewer must be installed by the client and cannot be loaded into the client browser by the server?  And it cannot automatically play ePubs loaded from the server?

And https://github.com/readium/web is not ready for release?  Where are the several Readium 2 cloud reader implementations that are working?  Are they under active development and support?  Can I download and install them on my CentOS server?

**Reply by danielweck** [reply]
readium-js-viewer; just like epub.js; does not require a server-side dynamic runtime, the reader app can simply be served by a static host (everything runs in the browser)

I will let other speak with regards to the Readium2 web reader projects.

**Reply by danielweck** [reply]
Here's a demo deployment of Readium1 web reader:

https://readium.web.app

**Reply by mickael-menu** [reply]
https://github.com/readium/web is still in its infancy but is expected to get significant development this year. For now there's only the shared models but no navigator (see [`architecture`](https://github.com/readium/architecture) for terminology).

**Reply by danielweck** [reply]
To install Readium1 web reader, download this file:
 Readium_cloud-reader.zip

...from this URL:
https://github.com/readium/readium-js-viewer/releases/tag/0.31.1

...and unzip in your static server folder.

**Reply by danielweck** [reply]
I see you have already outreached to one of the Readium2 adaptations:
https://github.com/d-i-t-a/R2D2BC/discussions/136

**Reply by danielweck** [reply]
Out of curiosity, I tried opening a publication from your collection, directly from a URL that references a packaged / zipped EPUB (i.e. not exploded / unzipped):
http://resources.aionianbible.org/Holy-Bible---English---World-English-Bible---Source-Edition.epub
=>
https://readium.web.app/?epub=http%3A%2F%2Fresources.aionianbible.org%2FHoly-Bible---English---World-English-Bible---Source-Edition.epub

This doesn't work because of HTTP CORS issues, I think. But it should work if the content is hosted on the same domain (aka origin) as your deployed reader app URL.

**Reply by danielweck** [reply]
An important point about loading packaged / zipped EPUBs from the Readium1 web reader is that unzipping occurs dynamically on the client side, and that is a computationally-expensive operation for a 5MB publication like the aforementioned one from your collection.
I would recommend extracting publication resources on the server and host them statically in their own dedicated folders.

**Reply by Nainoia-Inc-Admin** [reply]
Maybe readium-js-viewer is just what I want then.  Where can I go for instructions on installation and configuration?

**Reply by danielweck** [reply]
In the Readium2 architecture, there is a server component which can "streams" publication resources directly from packaged / zipped EPUBs. For example, with the aforementioned publication:
https://readium2.herokuapp.com/pub/aHR0cDovL3Jlc291cmNlcy5haW9uaWFuYmlibGUub3JnL0hvbHktQmlibGUtLS1FbmdsaXNoLS0tV29ybGQtRW5nbGlzaC1CaWJsZS0tLVNvdXJjZS1FZGl0aW9uLmVwdWI%3D/manifest.json/show/all

**Reply by danielweck** [reply]
Just like epub.js, the Readium1 web reader is capable of loading the special JSON "manifest" shown in the above comment.
For example: https://readium.web.app/?epub=https%3A%2F%2Freadium2.herokuapp.com%2Fpub%2FaHR0cDovL3Jlc291cmNlcy5haW9uaWFuYmlibGUub3JnL0hvbHktQmlibGUtLS1FbmdsaXNoLS0tV29ybGQtRW5nbGlzaC1CaWJsZS0tLVNvdXJjZS1FZGl0aW9uLmVwdWI%3D%2Fmanifest.json

...unfortunately this link doesn't work right now despite correct HTTP CORS headers (I haven't checked this for a while, so I'm not sure exactly what breaks)

I am away from my computer right now so cannot test thoroughly, sorry.

**Reply by danielweck** [reply]
Ah sorry, the link above was incorrect, this one works:

https://readium.web.app/?epub=https%3A%2F%2Freadium2.herokuapp.com%2Fpub%2FaHR0cDovL3Jlc291cmNlcy5haW9uaWFuYmlibGUub3JnL0hvbHktQmlibGUtLS1FbmdsaXNoLS0tV29ybGQtRW5nbGlzaC1CaWJsZS0tLVNvdXJjZS1FZGl0aW9uLmVwdWI%3D

**Reply by Nainoia-Inc-Admin** [reply]
hey that is very nice.  Yes is that the web streamer or the client side player like epub.js.  I assume it is the client side player.  

Is this the player dowloaded from https://github.com/readium/readium-js-viewer/releases/tag/0.31.1? 

And are configuration options explained in the download?

Also what are the pros-cons of the client side player vs the web player?  Though I think you are saying the web server side player doesn't work yet.

**Reply by danielweck** [reply]
The web reader deployed at https://readium.web.app was automatically uploaded from this GitHub "release":
https://github.com/readium/readium-js-viewer/releases/tag/0.32.0-alpha
(there is a continuous integration script which triggers a build each time changes are pushed to the code repository)

There is no significant difference in terms of end-user features between the above and this release:
https://github.com/readium/readium-js-viewer/releases/tag/0.31.1

**Reply by danielweck** [reply]
To install and test in your own server, simply download the file named `Readium_cloud-reader.zip` from one of the above GitHub releases (expand the "assets" section to see the file attachments), and unzip into a statically-served folder on your server (just like you do with epub.js).

**Reply by danielweck** [reply]
It is recommended to unzip your EPUB files into the `epub_content` folder of the Readium1 web reader. The `epub_library.json` file can be used to expose your publications in the bookshelf user interface. See for example:
https://github.com/readium/readium-js-viewer/blob/master/epub_content/epub_library.json

**Reply by Nainoia-Inc-Admin** [reply]
Okay I finally had the time to install Readium and it works great.  Thanks!  I have three questions...

1. In the viewer mode there is a button on the top right to jump to the Library.  How can I remove that button?  I maintain my own index to my epubs and don't want my users to use that button.

2. When in the Library there is also a button to "Add (upload) items to the library".  I want to remove that button also, just in case someone happens to get into the Library, I don't want them to upload anything or even use my server to view their own epubs.

3. I am a little nervous about that upload button.  Are there any known vulnerabilities in the Readium software that threaten the security of my server?  Any other options that I can turn off to make things more secure?

Here is a sample of Readium on my website

http://stage.aionianbible.org/Readium/Holy-Bible---English---World-English-Bible---Source-Edition 

That is a staging site so do not publish the link any further.

Thanks again!

**Reply by danielweck** [reply]
>     1. In the viewer mode there is a button on the top right to jump to the Library.  How can I remove that button?  I maintain my own index to my epubs and don't want my users to use that button.

Add this CSS snippet to your HTML `<style>` tag in the `<head>`:
```css
.btn.icon-library { display: none; }
```

**Reply by danielweck** [reply]
>     2\. When in the Library there is also a button to "Add (upload) items to the library".  I want to remove that button also, just in case someone happens to get into the Library, I don't want them to upload anything or even use my server to view their own epubs.

Add this CSS to your library HTML file:

```css
#empty-message { display: none; }
.btn-group.navbar-right { display: none; }
```

**Reply by danielweck** [reply]
>     3\. I am a little nervous about that upload button.  Are there any known vulnerabilities in the Readium software that threaten the security of my server?  Any other options that I can turn off to make things more secure?

This button actually does nothing at all :)
It was put there as a placeholder for a future possible feature, which never materialized.

**Reply by Nainoia-Inc-Admin** [reply]
Yes of course.  why didn't I think of that.  Thanks!

Your css needed a little tune up.  this worked for me.

#addbutt { display: none !important; }
#empty-message { display: none !important; }
#empty-message-arrow { display: none !important; }
#app-navbar .icon-library { display: none !important; }

**Reply by danielweck** [reply]
👍  :)
